Fix compilation errors in run-epanet and clean up warnings spewed by MSVC

ttaxon opened the pull request:
run-epanet was not being generated under MSVC due to EN_runEPANET not being
exported.  The CMake directive GenerateExportHeader was used to create the
header and epanet3.cpp and epanet3.h were modified as appropriate.

Also, for MSVC, the default warning level was changed from all to W4.  This
significantly reduced the number of "useless" warnings generated.  A few that
were disabled by default that I felt were useful were enabled.

The ones re-enabled are:
C4571: Related to exception handling
C4263 & C4264: related to virtual functions not being overridden "properly"

There also should likely be something done about exception handling.  The default settings for MSVC result in the following warnings:

warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc

If /EHsc is specified then the warning 4571 is issued:

Informational: catch(...) semantics changed since Visual C++ 7.1; structured exceptions (SEH) are no longer caught

[Here](https://msdn.microsoft.com/en-us/library/55s8esw4.aspx) is the link to the MSDN page talking about this warning and what to do about it.  This is beyond my level of expertise in C++.

ttaxon committed:
Fix compilation errors in run-epanet and clean up warnings spewed by MSVC
run-epanet was not being generated under MSVC due to EN_runEPANET not being
exported.  The CMake directive GenerateExportHeader was used to create the
header and epanet3.cpp and epanet3.h were modified as appropriate.

Also, for MSVC, the default warning level was changed from all to W4.  This
significantly reduced the number of "useless" warnings generated.  A few that
were disabled by default that I felt were usefule were enabled.

samhatchett commented:
Hi @ttaxon ! Looks like this breaks build on linux - missing header, which I guess should be generated by CMake? do you think that adding the `WINDOWS_EXPORT_ALL_SYMBOLS` define when calling cmake (like in #6 ) would do the same thing?

ttaxon committed:
Fix export include issue

Added -I. to non-MSVC generators so the epanet3_Export.h can be found

ttaxon commented:
Interesting...  I tested it on the mac command line and it worked fine.  I know its not linux, but it is close.  cmake created the epanet3_Export.h file.  I'm personally not a fan of exporting all symbols if it can be avoided.

I found the issue.  I added /I. to the MSVC specific flags, but since my Mac didn't need it, I assumed linux didn't either.  I have added -I. to the CXX_FLAGS for the non-MSVC generators.  Should I issue another pull request or send you the patch?

samhatchett commented:
Thanks Tom. If you commit to the branch where the PR is parked, then the request is automatically updated. See above, tests passed 🎉 

Do you have some literature to pass around about exporting symbols? I'm a little sad to have to clutter the declarations since no other platform needs that.

ttaxon commented:
Sam, it has been my preference/practice to just export only the symbols needed - I'm not sure of the origins of that.  Obviously on linux it is not needed.

I'll just add set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON) to the the MSVC portion of the CMakeLists.txt and remove the reference to the epanet3_Export.h and the export defined used in epanet3.cpp.  Sound like a plan?

samhatchett commented:
cool, just curious since i don't have much involvement in the windows world. obviously the option exists for a reason...

ttaxon committed:
Enable CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS for MSVC builds

Removed commands to create exort header file and also removed
epanet3_Export from EN_runEpanet definition and declaration.

ttaxon commented:
Should be good to go now.

Interesting aside...  I did a bit of searching to try to find out _why_ functions need to be exported from a DLL and I didn't easily find anything explaining the need.

samhatchett commented:
Hey @ttaxon thanks for looking into that! Hate to complicate matters, but would you mind looking at the README section on building for windows, and make sure that it still makes sense with this cmake configuration?

ttaxon committed:
Update README.md to reflect the changes made to Cmake for Viusual Studio

No longer need to specify CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS on the command
line as it is in the CMaakeLists.txt file now.

samhatchett commented:
great! thanks @ttaxon
